### ISSUE TITLE
Server can timeout client before time actually expires

### ISSUE THREAD

**Issue by igough** [eclipse/mosquitto#961]
In theory, the mosquitto server tracks whether a client times out when the current time is 1.5x the keep alive time after the last received message from that client. But the saved time in the server is a truncation of the current time to its value in seconds. For example, for linux, from mosquitto_time() in `lib/time_mosq.c`

```
struct timespec tp;

clock_gettime(CLOCK_MONOTONIC, &tp);
return tp.tv_sec;
```
which keeps the seconds, but throws away the nanoseconds.

From src/loop.c

```
now = mosquitto_time();
...
if(!(context->keepalive)
   || context->bridge
   || now - context->last_msg_in < (time_t)(context->keepalive)*3/2)
```
If you set the client keep alive to be 2 seconds, the timeout *should* be at 3 seconds, but if the last message arrived at 23.999 and the current time is 26.0000 when it checks to expire a client, the server thinks that 26-23=3 seconds means the client has timed out, when in reality, the elapsed time since the last message is 26.000-23.999=2.001 seconds, which force disconnects the client before it should.

Internally, the mosquitto server should save its times such as context->last_msg_in at a higher precision than 1 second to alleviate this problem. The mosquitto_time() function is already retrieving higher resolution time for all 3 platforms, but is throwing away the higher precision.

And I know you are going to ask, so, I need to use a short keepalive timeout because MQTT is used as the messaging system for a distributed robotics system. Setting the keepalive to 60 seconds means a robot out of control for 90 seconds in the case of an actual timeout. That is generally regarded as Not Good(TM). A keepalive of 2 means only 3 seconds of careening into people and things. Of course, with a higher internal time precision, a keepalive of 1.000 second would translate to a timeout of 1.500 seconds instead of a timeout of 1 second (since 1 second x 1.5, truncates to 1 second), which would be even better in loss of control situations.

**Comment by toast-uz** [comment]
The completely contrary request has been accepted recently. Refer #865 . How should we do? Managing milliseconds seems ineffective.

**Comment by toast-uz** [comment]
Sorry I've confused. This issue has already solved on #865 . Could you confirm the latest codes?

**Comment by igough** [comment]
Roger's logic would also solve the problem.
